evap/staff/views.py

```diff
@@ -311,27 +313,51 @@ def semester_import(request, semester_id):
     semester = get_object_or_404(Semester, id=semester_id)
     raise_permission_denied_if_archived(semester)
 
-    form = ImportForm(request.POST or None, request.FILES or None)
+    excel_form = ImportForm(request.POST or None, request.FILES or None)
 
-    if form.is_valid():
+    errors = []
+    warnings = []
+
+    if request.method == "POST":
         operation = request.POST.get('operation')
         if operation not in ('test', 'import'):
             raise SuspiciousOperation("Invalid POST operation")
 
-        # extract data from form
-        excel_file = form.cleaned_data['excel_file']
-        vote_start_date = form.cleaned_data['vote_start_date']
-        vote_end_date = form.cleaned_data['vote_end_date']
-
         test_run = operation == 'test'
-
-        # parse table
-        EnrollmentImporter.process(request, excel_file, semester, vote_start_date, vote_end_date, test_run)
-        if test_run:
-            return render(request, "staff_semester_import.html", dict(semester=semester, form=form))
-        return redirect('staff:semester_view', semester_id)
+        import_run = operation == 'import'
+
+        excel_form_valid = excel_form.is_valid()
+
+        if test_run and excel_form_valid:
+            excel_file = excel_form.cleaned_data['excel_file']
+            warnings, errors = EnrollmentImporter.process(request, excel_file, semester, None, None, test_run)
+            if not errors:
+                save_import_file(excel_file, str(request.user.id))
+                return render(request, "staff_semester_import.html", dict(semester=semester,
+                    excel_form=excel_form, warnings=warnings, errors=errors, test_passed=True))
+            else:
+                delete_import_file(str(request.user.id))
+                return render(request, "staff_semester_import.html", dict(semester=semester,
+                    excel_form=excel_form, warnings=warnings, errors=errors, test_passed=False))
+
+        elif import_run:
+            import_filename = get_import_file_name_if_exists(str(request.user.id))
+            if import_filename:
+                vote_start_date = excel_form.cleaned_data['vote_start_date']
+                vote_end_date = excel_form.cleaned_data['vote_end_date']
+                if vote_start_date is None or vote_end_date is None:
```

Comment:
please check if you can make these field required based on the action and do a form.is_valid instead of this here